### PR TITLE
GH-39398: [C++][Parquet] Use std::count in ColumnReader ReadLevels

### LINKED ISSUE #39398: [C++] [Parquet] Use std::count in parquet ColumnReader
### Describe the enhancement requested

I've found that for-loop here
https://github.com/apache/arrow/blob/7c3480e2f028f5881242f227f42155cf833efee7/cpp/src/parquet/column_reader.cc#L1055-L1073
transforms into

0xc0c2f0 <ReadLevels()+96>      inc    %rdx
0xc0c2f3 <ReadLevels()+99>      cmp    %rax,%rdx
0xc0c2f6 <ReadLevels()+102>     jge    0xc0c30c <ReadLevels()+124>
0xc0c2f8 <ReadLevels()+104>     cmp    %cx,(%r14,%rdx,2)
0xc0c2fd <ReadLevels()+109>     jne    0xc0c2f0 <ReadLevels()+96>
0xc0c2ff <ReadLevels()+111>     incq   0x0(%rbp)                                                   
0xc0c303 <ReadLevels()+115>     mov    (%rbx),%rax
0xc0c306 <ReadLevels()+118>     jmp    0xc0c2f0 <ReadLevels()+96>

That means that it uses iteration element by element and changes reference with incq
I think that the reason is that values_to_read and num_def_levels are not set as restrict. So the compiler can not optimize this to a more efficient way(for example using simd)

On my flamegraph this part showed ~10% of time spent

### Component(s)

C++, Parquet

### PR DESCRIPTION
**Pull request by Hattonuri**
### Rationale for this change

I've found that for-loop here
https://github.com/apache/arrow/blob/7c3480e2f028f5881242f227f42155cf833efee7/cpp/src/parquet/column_reader.cc#L1055-L1073

transforms into

0xc0c2f0 <ReadLevels()+96>      inc    %rdx
0xc0c2f3 <ReadLevels()+99>      cmp    %rax,%rdx
0xc0c2f6 <ReadLevels()+102>     jge    0xc0c30c <ReadLevels()+124>
0xc0c2f8 <ReadLevels()+104>     cmp    %cx,(%r14,%rdx,2)
0xc0c2fd <ReadLevels()+109>     jne    0xc0c2f0 <ReadLevels()+96>
0xc0c2ff <ReadLevels()+111>     incq   0x0(%rbp)                                                   
0xc0c303 <ReadLevels()+115>     mov    (%rbx),%rax
0xc0c306 <ReadLevels()+118>     jmp    0xc0c2f0 <ReadLevels()+96>

That means that it uses iteration element by element and changes reference with incq
I think that the reason is that values_to_read and num_def_levels are not set as restrict. So the compiler can not optimize this to a more efficient way(for example using simd)

On my flamegraph this part showed ~10% of time spent

In this file there also some for loops which could easily be changed to std::count, but they do not touch references and I don't know the reason why std::count was not used in the all cpp/src/parquet/ directory - so I didn't change much

### What changes are included in this PR?

Using `std::count` in `parquet/column_reader.cc` to avoid loop not being optimized

### Are these changes tested?
They are tested with unittest but not benched because I don't know what bench will show performance rise here(

### Are there any user-facing changes?

* Closes: #39398

### CODE REVIEW

**Comment by kou** [comment]
This is not a MINOR change: https://github.com/apache/arrow/blob/main/CONTRIBUTING.md#Minor-Fixes
Could you open a new issue for this?

**Comment by mapleFU** [comment]
The change looks good to me, I'll check some optimization options in compiler for this

out of curiousity, would you mind provide the compiler you're using?

**Comment by mapleFU** [comment]
> I think that the reason is that values_to_read and num_def_levels are not set as restrict. So the compiler can not optimize this to a more efficient way(for example using simd)

Maybe `max_def_level_` is also related

I'll approve it if you relate this patch to an issue

**Comment by Hattonuri** [comment]
> out of curiousity, would you mind provide the compiler you're using?

I use clang with libstdc++, -O3 -march=x86-64-v3 -fno-omit-frame-pointer -std=c++2a

File: cpp/src/parquet/column_reader.cc
Comment (mapleFU) [inline cpp/src/parquet/column_reader.cc:1197]: I'm ok with this but maybe this does not optimize, because `values_to_read` is a value here, the compiler might well optimize it ( since strict aliasing cannot mixing int16_t* and int64_t*

**Review by mapleFU — APPROVED** [review]
+1

I'll wait for a few days waiting for other review it after holiday

**Comment by kou** [comment]
Could you check the build failure on AppVeyor?

https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/48852924#L1696

```text
C:/projects/arrow/cpp/src/parquet/column_reader.cc(1195): error C2220: the following warning is treated as an error
C:/projects/arrow/cpp/src/parquet/column_reader.cc(1165): note: while compiling class template member function 'int64_t parquet::`anonymous-namespace'::TypedColumnReaderImpl<parquet::BooleanType>::ReadBatchSpaced(int64_t,int16_t *,int16_t *,bool *,uint8_t *,int64_t,int64_t *,int64_t *,int64_t *)'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\include\type_traits(325): note: see reference to class template instantiation 'parquet::`anonymous-namespace'::TypedColumnReaderImpl<parquet::BooleanType>' being compiled
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\include\memory(1440): note: see reference to class template instantiation 'std::is_convertible<_Yty *,_Ty *>' being compiled
        with
        [
            _Yty=parquet::`anonymous-namespace'::TypedColumnReaderImpl<parquet::BooleanType>,
            _Ty=parquet::ColumnReader
        ]
C:/projects/arrow/cpp/src/parquet/column_reader.cc(1290): note: see reference to class template instantiation 'std::_SP_pointer_compatible<parquet::`anonymous-namespace'::TypedColumnReaderImpl<parquet::BooleanType>,parquet::ColumnReader>' being compiled
C:/projects/arrow/cpp/src/parquet/column_reader.cc(1195): warning C4244: 'initializing': conversion from '__int64' to 'int', possible loss of data
```

**Comment by mapleFU** [comment]
Aha I think def_level is int16, but actually we need inc int64_t...

**Comment by mapleFU** [comment]
@ursabot please benchmark

**Comment by mapleFU** [comment]
```
diff --git a/cpp/src/parquet/column_reader.cc b/cpp/src/parquet/column_reader.cc
index a49e58afb..aa11dc57e 100644
--- a/cpp/src/parquet/column_reader.cc
+++ b/cpp/src/parquet/column_reader.cc
@@ -1059,10 +1059,11 @@ class TypedColumnReaderImpl : public TypedColumnReader<DType>,
 
     // If the field is required and non-repeated, there are no definition levels
     if (this->max_def_level_ > 0 && def_levels != nullptr) {
-      *num_def_levels = this->ReadDefinitionLevels(batch_size, def_levels);
+      int64_t num_def_levels_value = this->ReadDefinitionLevels(batch_size, def_levels);
       // TODO(wesm): this tallying of values-to-decode can be performed with better
       // cache-efficiency if fused with the level decoding.
-      for (int64_t i = 0; i < *num_def_levels; ++i) {
+      *num_def_levels = num_def_levels_value;
+      for (int64_t i = 0; i < num_def_levels_value; ++i) {
         if (def_levels[i] == this->max_def_level_) {
           ++(*values_to_read);
         }
@@ -1909,7 +1910,8 @@ class TypedRecordReader : public TypedColumnReaderImpl<DType>,
 
     // When reading dense we need to figure out number of values to read.
     const int16_t* def_levels = this->def_levels();
-    for (int64_t i = start_levels_position; i < levels_position_; ++i) {
+    int64_t levels_position = levels_position_;
+    for (int64_t i = start_levels_position; i < levels_position; ++i) {
       if (def_levels[i] == this->max_def_level_) {
         ++(*values_to_read);
       }
```

I think code like this might minimal, and I don't know if this could be ok in this case...

Also I don't know whether regression is related, will wait some days

**Comment by Hattonuri** [comment]
Does your code transforms into avx? Because you still write to *values_to_read

**Comment by mapleFU** [comment]
> Because you still write to *values_to_read

Emm since this is not a reduction, and compiler find there is no strict aliasing problem, I think it might not matter. We can waiting for pitrou back from holiday and take a look

**Comment by emkornfield** [comment]
It might be a benchmark coverage issue.  @Hattonuri if this is showing improvements on your own benchmarks, it might make sense to add more parquet benchmarks in a separate PR so that regressions aren't introduced afterwards.

**Comment by Hattonuri** [comment]
My benchmark is just a reader based on low-level parquet api reading a single dataset prefix. As it's written using static reflection - it does not use much time. And then i look on perf report and watch asm code of long functions

**Comment by mapleFU** [comment]
Maybe we can draft a read-level benchmark in parquet-column-reader-benchmark. in current `parquet-column-reader-benchmark` we can find some optimization, but it's a bit indirect

**Comment by mapleFU** [comment]
Hmmm @emkornfield `SKip` and `Read` in `parquet-column-reader` benchmark get faster in my MacOS, ReadLevels itself doesn't have any benchmark data

**Comment by emkornfield** [comment]
> Hmmm @emkornfield SKip and Read in parquet-column-reader benchmark get faster in my MacOS, ReadLevels itself doesn't have any benchmark directly, since it's a bit hacking to get it because it's private :-(

Maybe we should make ReadLevels public/visible for testing? Or at least for posterity, can we post the before after of the benchmark?

**Comment by mapleFU** [comment]
I'll draft a pr for benchmark only and posting the data there

**Comment by mapleFU** [comment]
@emkornfield I've create a draft pr: https://github.com/apache/arrow/pull/39486

In my M1 MacOS with Release (-O2)

Origin:

```
-------------------------------------------------------------------------------------------------------------
Benchmark                                                   Time             CPU   Iterations UserCounters...
-------------------------------------------------------------------------------------------------------------
ColumnReaderReadLevels/Repetition:1/BatchSize:1000    5309956 ns      5145464 ns          138 bytes_per_second=505.577M/s
ColumnReaderReadLevels/Repetition:2/BatchSize:1000    6481672 ns      6169167 ns          114 bytes_per_second=449.652M/s
```

New:

```
-------------------------------------------------------------------------------------------------------------
Benchmark                                                   Time             CPU   Iterations UserCounters...
-------------------------------------------------------------------------------------------------------------
ColumnReaderReadLevels/Repetition:1/BatchSize:1000     712139 ns       686560 ns         1087 bytes_per_second=3.70027G/s
ColumnReaderReadLevels/Repetition:2/BatchSize:1000    1705093 ns      1651455 ns          440 bytes_per_second=1.64035G/s
```

**Comment by Hattonuri** [comment]
@mapleFU  Could please you run your benchmark for https://github.com/apache/arrow/pull/39403?

**Comment by mapleFU** [comment]
This looks good to me, I will wait other committers to approve and merge it :-)

**Review by emkornfield — APPROVED** [review]

**Comment by pitrou** [comment]
This benchmarks shows a massive improvement:
https://conbench.ursa.dev/compare/benchmark-results/0658e4fc116778c480000601eaada951...06593d73f0097fa88000c545dd84021f/

Thank you @Hattonuri !

**Review by pitrou — APPROVED** [review]

**Pull request merged**
